src/wallet/rpcwallet.cpp

```diff
@@ -305,6 +305,199 @@ UniValue sendtoaddress(const UniValue& params, bool fHelp)
+        t_generated_dests.insert(item.first);
+    }
+
+    // Ensure we have every address that holds a balance. While this is likely to be redundant
+    // with respect to the entries in the address book for addresses generated by this wallet,
+    // there is not a guarantee that an externally generated address (such as one associated with
+    // a future unified incoming viewing key) will have been added to the address book.
+    for (const std::pair<CTxDestination, CAmount>& item : pwalletMain->GetAddressBalances()) {
+        auto script = GetScriptForDestination(item.first);
+        if (pwalletMain->HaveWatchOnly(script)) {
+            t_watchonly_dests.insert(item.first);
+        } else if (t_generated_dests.count(item.first) == 0) {
+            // assume that if we didn't add the address to the addrbook
+            // that it's a change address. Ideally we'd have a better way
+            // of checking this by exploring the transaction graph;
+            t_change_dests.insert(item.first);
+        } else {
+            // already accounted for in the address book
+        }
+    }
+
+    /// sprout addresses
+    std::set<libzcash::SproutPaymentAddress> sproutAddresses;
+    pwalletMain->GetSproutPaymentAddresses(sproutAddresses);
+
+    /// sapling addresses
+    std::set<libzcash::SaplingPaymentAddress> saplingAddresses;
+    pwalletMain->GetSaplingPaymentAddresses(saplingAddresses);
+
+    // legacy_random source
+    {
+        // Add legacy randomly generated address records to the result.
+        // This includes transparent addresses generated by the wallet via
+        // the keypool and Sprout addresses for which we have the
+        // spending key.
+        UniValue random_t_addrs(UniValue::VARR);
+        for (const CTxDestination& dest : t_generated_dests) {
+            random_t_addrs.push_back(keyIO.EncodeDestination(dest));
+        }
+
+        UniValue random_t_change_addrs(UniValue::VARR);
+        for (const CTxDestination& dest : t_change_dests) {
+            random_t_change_addrs.push_back(keyIO.EncodeDestination(dest));
+        }
+
+        UniValue random_sprout_addrs(UniValue::VARR);
+        for (const SproutPaymentAddress& addr : sproutAddresses) {
+            if (HaveSpendingKeyForPaymentAddress(pwalletMain)(addr)) {
+                random_sprout_addrs.push_back(keyIO.EncodePaymentAddress(addr));
+            }
+        }
+
+        /// keypool source only applies to transparent transactions
+        UniValue random_t(UniValue::VOBJ);
+        random_t.pushKV("addresses", random_t_addrs);
+        random_t.pushKV("change_addresses", random_t_change_addrs);
+
+        UniValue random_sprout(UniValue::VOBJ);
+        random_sprout.pushKV("addresses", random_sprout_addrs);
+
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "legacy_random");
+        entry.pushKV("transparent", random_t);
+        entry.pushKV("sprout", random_sprout);
+
+        ret.push_back(entry);
```

Comment:
We should only add the entry if it contains addresses (and similarly we should e.g. only add the `sprout` object if we have Sprout addresses.